module/applications/actor/base-sheet.mjs

```diff
@@ -1115,6 +1135,52 @@ export default class ActorSheet5e extends ActorSheet {
 
   /* -------------------------------------------- */
 
+  /**
+   * Adjust the preparation mode of a dropped spell depending on the drop location on the sheet.
+   * @param {object} itemData    The item data requested for creation. **Will be mutated.**
+   */
+  _onDropSpell(itemData) {
+    if ( !["npc", "character"].includes(this.document.type) ) return;
+
+    // Determine the section it is dropped on, if any.
+    const mode = this._event.target.closest(".spellbook-section")?.dataset ?? {};
+
+    // Determine the actor's spell slot progressions, if any.
+    const progs = Object.values(this.document.classes).reduce((acc, cls) => {
+      if ( cls.spellcasting?.type === "pact" ) acc.pact = true;
+      else if ( cls.spellcasting?.type === "leveled" ) acc.leveled = true;
+      return acc;
+    }, {pact: false, leveled: false});
+
+    // Case 1: Drop a cantrip.
+    if ( itemData.system.level == 0 ) {
+      if ( ["pact", "prepared"].includes(mode["preparation.mode"]) ) {
+        itemData.system.preparation.mode = "prepared";
+      } else if ( !mode["preparation.mode"] ) {
+        const isCaster = this.document.system.details.spellLevel || progs.pact || progs.leveled;
+        itemData.system.preparation.mode = isCaster ? "prepared" : "innate";
+      } else {
+        itemData.system.preparation.mode = mode["preparation.mode"];
+      }
+    }
+
+    // Case 2: Drop a leveled spell.
+    else {
+      if ( (mode.level == 0) || !mode["preparation.mode"] ) {
+        if ( this.document.type === "npc" ) {
+          const npcLevel = this.document.system.details.spellLevel;
+          itemData.system.preparation.mode = npcLevel ? "prepared" : "innate";
+        } else {
+          itemData.system.preparation.mode = progs.pact ? "pact" : progs.leveled ? "prepared" : "innate";
+        }
+      } else {
+        itemData.system.preparation.mode = mode["preparation.mode"];
+      }
+    }
```

Comment:
Linter isn't a big fan of `else` blocks that only contain another `if` statement, so this can be refactored up a level:

```suggestion
    // Case 2: Drop a leveled spell in a section without a mode.
    else if ( (mode.level === 0) || !mode["preparation.mode"] ) {
      if ( this.document.type === "npc" ) {
        itemData.system.preparation.mode = this.document.system.details.spellLevel ? "prepared" : "innate";
      } else {
        itemData.system.preparation.mode = progs.pact ? "pact" : progs.leveled ? "prepared" : "innate";
      }
    }

    // Case 3: Drop a leveled spell in a specific section.
    else {
      itemData.system.preparation.mode = mode["preparation.mode"];
    }
```